File: ab-cloud/verification/javascript/ab_cloud_verify_ru.js
// ==============================================================================
// AB-Cloud Verification Suite — JavaScript/Node.js (Russian)
// ==============================================================================
// Русская версия. Для двуязычной, используйте ab_cloud_verify.js
// ==============================================================================

'use strict';

const fs = require('fs');
const path = require('path');
const readline = require('readline');
const zlib = require('zlib');

// --- Русские сообщения --------------------------------------------------------

const MSG_RU = {
  header:       'Комплекс проверки AB-Cloud — JavaScript/Node.js',
  separator:    '──────────────────────────────────────────────────────',
  loading:      'Загрузка нулей из: %s',
  loaded:       'Загружено %d нулей из %s',
  obj1_title:   'Возражение 1: Сходимость b(N)',
  obj1_desc:    'b(N) = (1/N) * Σ|γ_k - γ̃_k|, точки Грама через W Ламберта',
  obj1_n:       'N',
  obj1_bN:      'b(N)',
  obj1_status:  'Статус',
  obj1_converge:'СХОДИТСЯ — b(N) → 0 подтверждает AB-Cloud',
  obj1_stable:  'СТАБИЛЬНО — b(N) ≈ 0, AB-Cloud согласуется',
  obj1_diverge: 'РАСХОДИТСЯ — b(N) ↛ 0, возражение подтверждено',
  obj2_title:   'Возражение 2: KS-тест интервалов GUE',
  obj2_desc:    's_k = (γ_{k+1}-γ_k)·log(γ_k/2π)/(2π), сравн. с p(s)=(πs/2)·exp(-πs²/4)',
  obj2_stat:    'D-статистика',
  obj2_pval:    'p-значение',
  obj2_result:  'Результат',
  obj2_pass:    'ПРОЙДЕНО — интервалы GUE подтверждены (p > 0.05)',
  obj2_fail:    'НЕ ПРОЙДЕНО — интервалы GUE отклонены (p ≤ 0.05)',
  obj3_title:   'Возражение 3: Наклон убывания при больших T',
  obj3_desc:    'Регрессия log|γ_k - γ̃_k| от log(γ_k), ожид. наклон ≈ -0.5',
  obj3_slope:   'Наклон',
  obj3_stderr:  'Стд. ошибка',
  obj3_target:  'Цель',
  obj3_result:  'Результат',
  obj3_pass:    'ПРОЙДЕНО — Наклон ≈ -0.5, убывание AB-Cloud подтверждено',
  obj3_fail:    'НЕ ПРОЙДЕНО — Наклон отклоняется от -0.5',
  no_data:      'ОШИБКА: Нули не загружены. Проверьте каталог данных.',
  done:         'Проверка завершена.',
};

// --- Функция W Ламберта (главная ветвь), метод Халли -------------------------

function lambertW0(x) {
  if (x === 0) return 0;
  let w = x > 1 ? Math.log(x) - Math.log(Math.log(x)) : x;
  for (let iter = 0; iter < 50; iter++) {
    const ew = Math.exp(w), f = w * ew - x, fp = ew * (1 + w), fpp = ew * (2 + w);
    w -= (2 * f * fp) / (2 * fp * fp - f * fpp);
    if (Math.abs(f) < 1e-12 * Math.abs(x + 1)) break;
  }
  return w;
}

// --- Вычисление точки Грама через W Ламберта ---------------------------------

function gramPoint(n) {
  if (n <= 0) return 0;
  const TWO_PI = 2 * Math.PI;
  let g = TWO_PI * n / lambertW0(n / Math.E);
  for (let iter = 0; iter < 3; iter++) {
    const theta = 0.5 * g * Math.log(g / TWO_PI) - 0.5 * g - Math.PI / 8;
    const dtheta = 0.5 * Math.log(g / TWO_PI);
    g += (Math.PI * n - theta) / dtheta;
  }
  return g;
}

function gramPointsVec(nVec) { return nVec.map(n => gramPoint(n)); }

// --- Функция распределения интервалов GUE (гипотеза Вигнера) -----------------

function gueCDF(s) { return 1 - Math.exp(-Math.PI * s * s / 4); }

// --- Загрузка нулей из файлов данных (потоковая, поддерживает .gz) -----------

const FILE_MAP = {
  zeta_zeros_50000:     'zeta_zeros_50000.txt',
  zeta_zeros_500k:      'zeta_zeros_500k_odlyzko.txt',
  zeta_zeros_2M:        'zeta_zeros_2M_odlyzko.txt',
  zeta_zeros_highT:     'zeta_zeros_highT_blocks.txt',
  zeros6:               'zeros6.txt',
  zeta_zeros_50000_csv: 'zeta_zeros_50000.csv',
};

async function loadZeros(dataDir, count = 0, source = 'auto') {
  let selectedFile = null;
  if (source !== 'auto' && FILE_MAP[source]) {
    selectedFile = path.join(dataDir, FILE_MAP[source]);
  } else {
    for (const nm of Object.keys(FILE_MAP)) {
      const fp = path.join(dataDir, FILE_MAP[nm]);
      if (fs.existsSync(fp)) {
        selectedFile = fp;
        if (count > 0 && count <= 50000 && nm === 'zeta_zeros_50000') break;
      }
    }
  }
  if (!selectedFile || !fs.existsSync(selectedFile)) {
    throw new Error(`Файл данных не найден в: ${dataDir}`);
  }
  console.log(`Загрузка нулей из: ${selectedFile}`);

  const zeros = [];
  const isGz = selectedFile.endsWith('.gz');
  const isCsv = path.extname(selectedFile).toLowerCase() === '.csv';

  let stream = fs.createReadStream(selectedFile, { encoding: 'utf8' });
  if (isGz) {
    stream = fs.createReadStream(selectedFile).pipe(zlib.createGunzip()).setEncoding('utf8');
  }
  const rl = readline.createInterface({ input: stream, crlfDelay: Infinity });

  for await (const line of rl) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    if (isCsv) {
      const val = parseFloat(trimmed.split(',')[0]);
      if (!isNaN(val) && isFinite(val)) zeros.push(val);
    } else {
      for (const tok of trimmed.split(/\s+/).filter(t => t.length > 0)) {
        const val = parseFloat(tok);
        if (!isNaN(val) && isFinite(val)) zeros.push(val);
      }
    }
    if (count > 0 && zeros.length >= count) { rl.close(); break; }
  }
  if (count > 0 && zeros.length > count) zeros.length = count;
  console.log(`Загружено ${zeros.length} нулей из ${path.basename(selectedFile)}`);
  return zeros;
}

// --- Вспомогательные функции форматирования -----------------------------------

function padL(str, len) { return String(str).padStart(len); }
function padR(str, len) { return String(str).padEnd(len); }

// --- Возражение 1: Сходимость b(N) -------------------------------------------

function objection1(zeros) {
  const M = MSG_RU, N = zeros.length;
  console.log(`\n${M.separator}\n${M.obj1_title}\n${M.obj1_desc}\n${M.separator}`);
  const cp = [100,500,1000,5000,10000,50000,100000,500000,1000000].filter(c => c <= N);
  console.log(`\n${padR(M.obj1_n,10)}  ${padR(M.obj1_bN,14)}  ${padR(M.obj1_status,8)}`);
  console.log(`${padR('──────────',10)}  ${padR('──────────────',14)}  ${padR('────────',8)}`);
  const results = []; let prev = NaN;
  for (const c of cp) {
    const idx = Array.from({length:c},(_,i)=>i+1);
    const gram = gramPointsVec(idx);
    let sum = 0; for (let i=0;i<c;i++) sum += Math.abs(zeros[i]-gram[i]);
    const bN = sum/c;
    const st = isNaN(prev) ? '—' : bN < prev*1.05 ? '↓' : '↑';
    console.log(`${padL(c,10)}  ${padL(bN.toFixed(8),14)}  ${padR(st,8)}`);
    results.push({N:c,bN,status:st}); prev = bN;
  }
  const fb = results[results.length-1].bN;
  const v = fb<0.01 ? M.obj1_converge : fb<0.5 ? M.obj1_stable : M.obj1_diverge;
  console.log(`\n${v}`);
  return results;
}

// --- Возражение 2: KS-тест интервалов GUE ------------------------------------

function objection2(zeros) {
  const M = MSG_RU, N = zeros.length;
  console.log(`\n${M.separator}\n${M.obj2_title}\n${M.obj2_desc}\n${M.separator}`);
  const sp = [];
  for (let k=0;k<N-1;k++) {
    const d = (zeros[k+1]-zeros[k]) * Math.log(zeros[k]/(2*Math.PI)) / (2*Math.PI);
    if (isFinite(d)&&d>0) sp.push(d);
  }
  sp.sort((a,b)=>a-b);
  const n = sp.length;
  let Dp=0, Dm=0;
  for (let i=0;i<n;i++) {
    Dp = Math.max(Dp, (i+1)/n - gueCDF(sp[i]));
    Dm = Math.max(Dm, gueCDF(sp[i]) - i/n);
  }
  const D = Math.max(Dp,Dm);
  const lam = (Math.sqrt(n)+0.12+0.11/Math.sqrt(n))*D;
  let pv = 0;
  for (let k=-5;k<=5;k++) pv += Math.pow(-1,k)*Math.exp(-2*k*k*lam*lam);
  pv = Math.max(0,Math.min(1,pv));
  console.log(`\n${padR(M.obj2_stat+':',18)} ${D.toFixed(8)}`);
  console.log(`${padR(M.obj2_pval+':',18)} ${pv.toExponential(6)}`);
  console.log(`\n${pv>0.05 ? M.obj2_pass : M.obj2_fail}`);
  return {D, pValue:pv};
}

// --- Возражение 3: Наклон убывания при больших T -----------------------------

function objection3(zeros) {
  const M = MSG_RU, N = zeros.length;
  console.log(`\n${M.separator}\n${M.obj3_title}\n${M.obj3_desc}\n${M.separator}`);
  const start = Math.max(1,Math.floor(N*0.5));
  const idx = []; for (let i=start;i<=N;i++) idx.push(i);
  const gram = gramPointsVec(idx);
  const lg=[], ld=[];
  for (let i=0;i<idx.length;i++) {
    const dev = Math.abs(zeros[idx[i]-1]-gram[i]);
    if (isFinite(dev)&&dev>0) { lg.push(Math.log(zeros[idx[i]-1])); ld.push(Math.log(dev)); }
  }
  const n=lg.length;
  let sx=0,sy=0,sxy=0,sx2=0;
  for (let i=0;i<n;i++) { sx+=lg[i]; sy+=ld[i]; sxy+=lg[i]*ld[i]; sx2+=lg[i]*lg[i]; }
  const den=n*sx2-sx*sx;
  const a=(sy*sx2-sx*sxy)/den, slope=(n*sxy-sx*sy)/den;
  let ssr=0; for (let i=0;i<n;i++) ssr+=Math.pow(ld[i]-(a+slope*lg[i]),2);
  const se=Math.sqrt(ssr/(n-2)*n/den);
  console.log(`\n${padR(M.obj3_slope+':',18)} ${slope.toFixed(6)}`);
  console.log(`${padR(M.obj3_stderr+':',18)} ${se.toFixed(6)}`);
  console.log(`${padR(M.obj3_target+':',18)} -0.5`);
  console.log(`\n${Math.abs(slope-(-0.5))<0.15 ? M.obj3_pass : M.obj3_fail}`);
  return {slope, stderr:se};
}

// --- Главная функция (только русский) ----------------------------------------

async function abCloudVerifyRu(options = {}) {
  const { dataDir='../data', zeros=0, source='auto', objection='all' } = options;
  const M = MSG_RU;
  console.log(`\n${M.separator}\n${M.header}\n${M.separator}`);
  const gammas = await loadZeros(dataDir, zeros, source);
  if (gammas.length===0) { console.log(M.no_data); return null; }
  const results = {};
  if (['all','1'].includes(objection)) results.obj1 = objection1(gammas);
  if (['all','2'].includes(objection)) results.obj2 = objection2(gammas);
  if (['all','3'].includes(objection)) results.obj3 = objection3(gammas);
  console.log(`\n${M.separator}\n${M.done}\n`);
  return results;
}

module.exports = { abCloudVerifyRu, loadZeros, objection1, objection2, objection3, MSG_RU };

File: ab-cloud/verification/r/ab_cloud_verify_ru.R
# ==============================================================================
# AB-Cloud Verification Suite — R Implementation (Russian)
# ==============================================================================
# Русская версия. Для двуязычной, используйте ab_cloud_verify.R
# ==============================================================================

MSG_RU <- list(
  header       = "Комплекс проверки AB-Cloud — R",
  separator    = "──────────────────────────────────────────────────────",
  loading      = "Загрузка нулей из: %s",
  loaded       = "Загружено %d нулей из %s",
  obj1_title   = "Возражение 1: Сходимость b(N)",
  obj1_desc    = "b(N) = (1/N) * Σ|γ_k - γ̃_k|, точки Грама через W Ламберта",
  obj1_n       = "N",
  obj1_bN      = "b(N)",
  obj1_status  = "Статус",
  obj1_converge = "СХОДИТСЯ — b(N) → 0 подтверждает AB-Cloud",
  obj1_stable  = "СТАБИЛЬНО — b(N) ≈ 0, AB-Cloud согласуется",
  obj1_diverge = "РАСХОДИТСЯ — b(N) ↛ 0, возражение подтверждено",
  obj2_title   = "Возражение 2: KS-тест интервалов GUE",
  obj2_desc    = "s_k = (γ_{k+1}-γ_k)·log(γ_k/2π)/(2π), сравн. с p(s)=(πs/2)·exp(-πs²/4)",
  obj2_stat    = "D-статистика",
  obj2_pval    = "p-значение",
  obj2_result  = "Результат",
  obj2_pass    = "ПРОЙДЕНО — интервалы GUE подтверждены (p > 0.05)",
  obj2_fail    = "НЕ ПРОЙДЕНО — интервалы GUE отклонены (p ≤ 0.05)",
  obj3_title   = "Возражение 3: Наклон убывания при больших T",
  obj3_desc    = "Регрессия log|γ_k - γ̃_k| от log(γ_k), ожид. наклон ≈ -0.5",
  obj3_slope   = "Наклон",
  obj3_stderr  = "Стд. ошибка",
  obj3_target  = "Цель",
  obj3_result  = "Результат",
  obj3_pass    = "ПРОЙДЕНО — Наклон ≈ -0.5, убывание AB-Cloud подтверждено",
  obj3_fail    = "НЕ ПРОЙДЕНО — Наклон отклоняется от -0.5",
  no_data      = "ОШИБКА: Нули не загружены. Проверьте каталог данных.",
  done         = "Проверка завершена."
)

# --- Функция W Ламберта (главная ветвь), метод Халли ------------------------

lambert_W0_ru <- function(x, tol = 1e-12, max_iter = 50) {
  if (x == 0) return(0)
  w <- if (x > 1) log(x) - log(log(x)) else x
  for (i in seq_len(max_iter)) {
    ew  <- exp(w)
    f   <- w * ew - x
    fp  <- ew * (1 + w)
    fpp <- ew * (2 + w)
    w   <- w - (2 * f * fp) / (2 * fp * fp - f * fpp)
    if (abs(f) < tol * abs(x + 1)) break
  }
  return(w)
}

# --- Вычисление точки Грама через W Ламберта ---------------------------------

gram_point_ru <- function(n) {
  if (n <= 0) return(0)
  val <- 2 * pi * n / lambert_W0_ru(n / exp(1))
  for (iter in 1:3) {
    theta  <- 0.5 * val * log(val / (2 * pi)) - 0.5 * val - pi / 8
    dtheta <- 0.5 * log(val / (2 * pi))
    val    <- val + (pi * n - theta) / dtheta
  }
  return(val)
}

gram_points_vec_ru <- function(n_vec) sapply(n_vec, gram_point_ru)

# --- Функция распределения интервалов GUE (гипотеза Вигнера) -----------------

gue_cdf_ru <- function(s) 1 - exp(-pi * s^2 / 4)

# --- Загрузка нулей из файлов данных -----------------------------------------

load_zeros_ru <- function(dataDir, count = 0, source = "auto") {
  files <- list(
    zeta_zeros_50000      = file.path(dataDir, "zeta_zeros_50000.txt"),
    zeta_zeros_500k       = file.path(dataDir, "zeta_zeros_500k_odlyzko.txt"),
    zeta_zeros_2M         = file.path(dataDir, "zeta_zeros_2M_odlyzko.txt"),
    zeta_zeros_highT      = file.path(dataDir, "zeta_zeros_highT_blocks.txt"),
    zeros6                = file.path(dataDir, "zeros6.txt"),
    zeta_zeros_50000_csv  = file.path(dataDir, "zeta_zeros_50000.csv")
  )
  if (source != "auto" && source %in% names(files)) {
    selected <- files[[source]]
  } else {
    selected <- NULL
    for (nm in names(files)) {
      if (file.exists(files[[nm]])) {
        selected <- files[[nm]]
        if (count > 0 && count <= 50000 && nm == "zeta_zeros_50000") break
      }
    }
  }
  if (is.null(selected) || !file.exists(selected))
    stop("Файл данных не найден в: ", dataDir)

  cat(sprintf(MSG_RU$loading, selected), "\n")
  ext <- tools::file_ext(selected)
  zeros <- if (ext == "csv") {
    df <- read.csv(selected, comment.char = "#", header = FALSE)
    as.numeric(df[[1]])
  } else {
    lines <- readLines(selected, warn = FALSE)
    lines <- lines[!grepl("^\\s*#", lines) & nzchar(trimws(lines))]
    vals  <- as.numeric(trimws(lines))
    vals[!is.na(vals)]
  }
  if (count > 0 && count < length(zeros)) zeros <- zeros[1:count]
  cat(sprintf(MSG_RU$loaded, length(zeros), basename(selected)), "\n")
  return(zeros)
}

# --- Возражение 1: Сходимость b(N) -------------------------------------------

objection_1_ru <- function(zeros) {
  M <- MSG_RU; N <- length(zeros)
  cat("\n", M$separator, "\n", M$obj1_title, "\n", M$obj1_desc, "\n", M$separator, "\n\n")
  checkpoints <- c(100, 500, 1000, 5000, 10000, 50000, 100000, 500000, 1000000)
  checkpoints <- checkpoints[checkpoints <= N]
  results <- data.frame(N = integer(), bN = numeric(), Status = character(), stringsAsFactors = FALSE)
  prev_bN <- NA
  for (cp in checkpoints) {
    idx  <- 1:cp
    gram <- gram_points_vec_ru(idx)
    bN   <- mean(abs(zeros[idx] - gram))
    status <- if (is.na(prev_bN)) "—" else if (bN < prev_bN * 1.05) "↓" else "↑"
    results <- rbind(results, data.frame(N = cp, bN = bN, Status = status, stringsAsFactors = FALSE))
    prev_bN <- bN
  }
  cat(sprintf("%10s  %14s  %8s\n", M$obj1_n, M$obj1_bN, M$obj1_status))
  cat(sprintf("%10s  %14s  %8s\n", "──────────", "──────────────", "────────"))
  for (i in seq_len(nrow(results)))
    cat(sprintf("%10d  %14.8f  %8s\n", results$N[i], results$bN[i], results$Status[i]))
  final_bN <- results$bN[nrow(results)]
  verdict <- if (final_bN < 0.01) M$obj1_converge else if (final_bN < 0.5) M$obj1_stable else M$obj1_diverge
  cat("\n", verdict, "\n")
  return(results)
}

# --- Возражение 2: KS-тест интервалов GUE ------------------------------------

objection_2_ru <- function(zeros) {
  M <- MSG_RU; N <- length(zeros)
  cat("\n", M$separator, "\n", M$obj2_title, "\n", M$obj2_desc, "\n", M$separator, "\n\n")
  gammas  <- zeros[1:(N - 1)]
  deltas  <- zeros[2:N] - zeros[1:(N - 1)]
  log_fac <- log(gammas / (2 * pi)) / (2 * pi)
  s       <- deltas * log_fac
  s       <- s[is.finite(s) & s > 0]
  ks_res  <- ks.test(s, gue_cdf_ru)
  cat(sprintf("%18s: %.8f\n", M$obj2_stat, ks_res$statistic))
  cat(sprintf("%18s: %.6e\n", M$obj2_pval, ks_res$p.value))
  verdict <- if (ks_res$p.value > 0.05) M$obj2_pass else M$obj2_fail
  cat("\n", verdict, "\n")
  return(list(D = ks_res$statistic, p.value = ks_res$p.value))
}

# --- Возражение 3: Наклон убывания при больших T -----------------------------

objection_3_ru <- function(zeros) {
  M <- MSG_RU; N <- length(zeros)
  cat("\n", M$separator, "\n", M$obj3_title, "\n", M$obj3_desc, "\n", M$separator, "\n\n")
  start <- max(1, floor(N * 0.5))
  idx   <- start:N
  gram  <- gram_points_vec_ru(idx)
  dev   <- abs(zeros[idx] - gram)
  valid <- is.finite(dev) & dev > 0
  log_gamma <- log(zeros[idx][valid])
  log_dev   <- log(dev[valid])
  fit    <- lm(log_dev ~ log_gamma)
  slope  <- coef(fit)[2]
  stderr <- summary(fit)$coefficients[2, 2]
  cat(sprintf("%18s: %.6f\n", M$obj3_slope, slope))
  cat(sprintf("%18s: %.6f\n", M$obj3_stderr, stderr))
  cat(sprintf("%18s: -0.5\n", M$obj3_target))
  verdict <- if (abs(slope - (-0.5)) < 0.15) M$obj3_pass else M$obj3_fail
  cat("\n", verdict, "\n")
  return(list(slope = slope, stderr = stderr))
}

# --- Главная функция (только русский) ----------------------------------------

ab_cloud_verify_ru <- function(dataDir = "../data", zeros = 0, source = "auto",
                                objection = "all") {
  M <- MSG_RU
  cat("\n", M$separator, "\n", M$header, "\n", M$separator, "\n")
  gammas <- load_zeros_ru(dataDir, count = zeros, source = source)
  if (length(gammas) == 0) { cat(M$no_data, "\n"); return(invisible(NULL)) }
  results <- list()
  if (objection %in% c("all", "1")) results$obj1 <- objection_1_ru(gammas)
  if (objection %in% c("all", "2")) results$obj2 <- objection_2_ru(gammas)
  if (objection %in% c("all", "3")) results$obj3 <- objection_3_ru(gammas)
  cat("\n", M$separator, "\n", M$done, "\n\n")
  return(invisible(results))
}
